diesel/src/connection/commit_error_processor.rs

```diff
@@ -0,0 +1,64 @@
+//! A module to evaluate what to do when a commit triggers an error.
+use crate::result::{DatabaseErrorKind, Error};
+
+use super::ValidTransactionManagerStatus;
+
+/// Transaction status returned upon error on commit.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum CommitErrorOutcome {
+    /// Necessitates a rollback to return to a valid transaction
+    RollbackAndThrow(Error),
+    /// Broken transaction. Returned if an error has occurred earlier in a Postgres transaction.
+    Throw(Error),
+    /// Broken transaction. Similar to `Throw`, but marks the manager as broken. It should switch
+    /// to `TransactionManagerStatus::InError` and refuse to run additional operations.
+    ThrowAndMarkManagerAsBroken(Error),
+}
+
+/// Trait needed for the transaction manager.
+pub trait CommitErrorProcessor {
+    /// Returns the status of the transaction following an error upon commit.
+    /// When any of these kinds of error happen on `COMMIT`, it is expected
+    /// that a `ROLLBACK` would succeed, leaving the transaction in a non-broken state.
+    /// If there are other such errors, it is fine to add them here.
+    fn process_commit_error(&self, error: Error) -> CommitErrorOutcome;
+}
+
+/// Default implementation of CommitErrorProcessor::process_commit_error(), used for MySql and
+/// Sqlite connections. Returns `CommitErrorOutcome::RollbackAndThrow` if the transaction depth is
+/// greater than 1, the error is a `DatabaseError` and the error kind is either
+/// `DatabaseErrorKind::SerializationFailure` or `DatabaseErrorKind::ReadOnlyTransaction`
+pub fn default_process_commit_error(
+    transaction_state: &ValidTransactionManagerStatus,
+    error: Error,
+) -> CommitErrorOutcome {
+    let transaction_depth = transaction_state.transaction_depth();
+    match error {
+        Error::DatabaseError(DatabaseErrorKind::ReadOnlyTransaction, _)
+        | Error::DatabaseError(DatabaseErrorKind::SerializationFailure, _)
+            if transaction_depth.is_none() =>
```

Comment:
After thinking again about this I believe that's not correct. 
```suggestion
            if transaction_depth == NonZeroU32::new(1) =>
```

(Will fix that next week + add an test for this case)